src/middlewared/middlewared/plugins/interface/listen.py

```diff
@@ -0,0 +1,103 @@
+import asyncio
+from collections import namedtuple
+
+from middlewared.service import accepts, Service, private
+
+PreparedDelegate = namedtuple("PreparedDelegate", ["delegate", "state", "addresses"])
+
+PRE_SYNC_LISTEN_DELEGATES = []
+POST_ROLLBACK_LISTEN_DELEGATES = []
+
+
+class InterfaceService(Service):
+
+    delegates = []
+
+    class Config:
+        namespace_alias = "interfaces"
+
+    @private
+    def register_listen_delegate(self, delegate):
+        self.delegates.append(delegate)
+
+    @accepts()
+    async def services_restarted_on_sync(self):
+        """
+        Returns which services will be set to listen on 0.0.0.0 (and, thus, restarted) on sync.
+
+        Example result:
+        [
+            // Samba service will be set ot listen on 0.0.0.0 and restarted because it was set up to listen on
+            // 192.168.0.1 which is being removed.
+            {"type": "system-service", "service": "cifs", "ips": ["192.168.0.1"]},
```

Comment:
Very minor nit, wonder if we should make that uppercase, just to be consistent with some other places, since its a kind of ENUM